### DISCUSSION TITLE
Similiar Images

### DISCUSSION THREAD

**Discussion by MoxBe** [ImageMagick/ImageMagick#6330] [Help]
### ImageMagick version

7.0.8-56

### Operating system, version and so on

Windows 7

### Description

I need help how to compare similar images.
Under are three .png images that I want to compare.
How to do it with ImageMagick?
1.png and 2.png are similar or equal, while others are not similar.
First, I resize images so that they are equal size and then compare.
I tried to use compare -metric NCC option to find out if they are similar, but I don't understand results.
Can someone help me?

![1](https://github.com/ImageMagick/ImageMagick/assets/53977638/c107b2c8-f066-46a3-8272-a4c8fc734112)
![2](https://github.com/ImageMagick/ImageMagick/assets/53977638/5fe5d371-4c1e-4bf7-89d2-e6c936322623)
![3](https://github.com/ImageMagick/ImageMagick/assets/53977638/2577740d-7b95-4dd2-9bce-39ae6b934bed)

**Comment by fmw42** [comment]
The first number is the similarity in raw values in range 0 to quantum range. The value in parentheses is the similarity in the range 0 to 1.  So multiply by 100 to get 0 to 100% similarity for NCC.

**Reply by MoxBe** [reply]
I don't get values that you described.
These are the values that I get:

compare -metric NCC 1.png 1.png diff.png
105.672

compare -metric NCC 1.png 2.png diff.png
6.44662e+018

compare -metric NCC 1.png 3.gif diff.png
3.29637

**Comment by urban-warrior** [comment]
We get expected results from the latest release of ImageMagick.   You may need to upgrade your release.

**Reply by MoxBe** [reply]
OK, I have installed latest release, and it returns expected results, but I need an advice.
I have two images:
![1](https://github.com/ImageMagick/ImageMagick/assets/53977638/ff94daa9-1805-46df-b1d8-c492f36d2b15)
![2](https://github.com/ImageMagick/ImageMagick/assets/53977638/1c0d351c-5bf5-4e2b-8af3-59d9c187c1e4)


Compare results are:
compare -metric NCC 1.png 2.png diff.png
0.335823

Although the images are quite similar,, almost identical,  I get 33% similarity.
Why only 33% if diff shows just border is different?
![diff](https://github.com/ImageMagick/ImageMagick/assets/53977638/b9a75945-0121-4c85-abee-deb80d074288)

What I need is to compare images for general shape similarity, not by colors.

**Reply by MoxBe** [reply]
I think I know why only 33% similarity.

White fill of both images is actually transparent, but it can't be seen here.
When I filled transparent with white color, result was 98%.

**Comment by MoxBe** [comment]
I don't quite understand this.
I have two images:
![1](https://github.com/ImageMagick/ImageMagick/assets/53977638/03e66b14-e98f-4ea1-ae7f-cb64e772ec51)
![2](https://github.com/ImageMagick/ImageMagick/assets/53977638/94f3938b-8b8f-49f3-8075-7906e26f3a98)

They are both transaprent (it's not obvious here).
I convert them both to be grayscale, without alpha and same size.

convert 1.png -background white -alpha remove -alpha off -set colorspace Gray -separate -average a.png
convert 2.png -background white -alpha remove -alpha off -set colorspace Gray -separate -average -resize 505x84! b.png

They look like this:
![a](https://github.com/ImageMagick/ImageMagick/assets/53977638/ed56e60b-0a5e-4386-9c9c-c26c8470c191)
![b](https://github.com/ImageMagick/ImageMagick/assets/53977638/2fd29c9d-ee7b-4ec1-8e89-a3dca5eca9dc)

Now I do compare:
compare -metric NCC a.png b.png diff.png

and I get result 0.851389

Why?

If I do 
compare -metric AE a.png b.png diff.png
I get absolute pixels difference is 2412 which is 5% of total image pixels, meaning that similarity should be 95% and not 85% as magick comparison result showed.

I experiment a little bit, and found out that maybe resize is the issue.
If I do resize of 2.png in Paint Shop Pro and compare it, then I get 95% similarity.

Can someone help me about this?

**Comment by snibgo** [comment]
IM provides many metrics for similarity. There is no single "correct" value. See https://imagemagick.org/script/command-line-options.php#metric

I generally use RMSE. For your images this gives a value of 0.0356294 on a scale of 0.0 to 1.0, where 0.0 means the images are identical.  I generally reckon that where the value is less than 0.01, we can't visually see a difference.

**Comment by MoxBe** [comment]
I'll try RMSE, but I am not sure if metric type is an issue here.
As I explained, resizing image in magick and Paint Shop Pro gives big difference in similarity, although they look the same when viewing with a naked eye. 
Images are so simple that there shouldn't be issues at all.

**Comment by fmw42** [comment]
Please post your two input images.

**Comment by MoxBe** [comment]
I already did, look at the previous posts.

**Reply by fmw42** [reply]
I only see two white rectangles.  Use RMSE for such images.  

```
magick compare -metric rmse 238341559-ed56e60b-0a5e-4386-9c9c-c26c8470c191.png 238341622-2fd29c9d-ee7b-4ec1-8e89-a3dca5eca9dc.png null:
2334.97 (0.0356294)
```

This is about 3.6% differences, which is reasable in my opinion for the visual differences that I see.
